File: algorithm_questions/two_sum.ex
defmodule Solution do
  @spec two_sum(nums :: [integer], target :: integer) :: [integer]
    def two_sum(nums, target) do

        # v1
        # pairs = for {x, i} <- Enum.with_index(nums),
        #     {y, j} <- Enum.with_index(nums),
        #     i < j,
        #     x + y == target,
        #     do: [i, j]
        # List.first(pairs)

        # v2
        # nums
        # |> Enum.with_index()
        # |> Enum.reduce_while(%{}, fn {num, idx}, acc -> 
        #     need = target - num
        #     if Map.has_key?(acc, need) do
        #         {:halt, [acc[need], idx]}
        #     else
        #         {:cont, Map.put(acc, num, idx)}
        #     end
        # end)
    end
end

File: algorithm_questions/two_sum.ts
// brute force
function twoSumB(nums: number[], target: number): [number, number] | null {
	const n = nums.length;
	for (let i=0; i<n; i++) {
		for (let j=1; i<n; j++) {
			if (nums[i]+nums[j] === target) {
				return [i, j];
			}
		}
	}
	return null;
}

// map
function twoSumH(nums: number[], target: number): [number, number] | null {
	const seen: Record<number, number> = {};

	for (let i=0; i<nums.length; i++) {
		const num = nums[i];
		const need = target - num;
		if (need in seen) {
			return [seen[need], i];
		}
		seen[num]=i;
	}
	return null;
}
